src/main/java/pl/datamatica/traccar/api/providers/RouteProvider.java

```diff
@@ -0,0 +1,226 @@
+            + " left join fetch r.corridor";
+    
+    private final User requestUser;
+    private DeviceProvider devices;
+    private GeoFenceProvider geofences;
+    
+    public RouteProvider(EntityManager em, User requestUser) {
+        super(em);
+        this.requestUser = requestUser;
+    }
+    
+    public void setDeviceProvider(DeviceProvider devices) {
+        this.devices = devices;
+    }
+    
+    public void setGeofenceProvider(GeoFenceProvider geofences) {
+        this.geofences = geofences;
+    }
+    
+    public Stream<DbRoute> getAllAvailableRoutes(boolean archive) throws ProviderException {
+        if(!requestUser.hasPermission(UserPermission.TRACK_READ))
+            throw new ProviderException(Type.ACCESS_DENIED);
+        if(requestUser.hasPermission(UserPermission.ALL_TRACKS))
+            return em.createQuery(SELECT_WITH_JOINS+" where r.archive = :archive", DbRoute.class)
+                    .setParameter("archive", archive).getResultList().stream();
+        return em.createQuery(SELECT_WITH_JOINS+" where r.owner = :user and r.archive = :archive", DbRoute.class)
+                .setParameter("user", requestUser).setParameter("archive", archive)
+                .getResultList().stream();
+    }
+    
+    public DbRoute updateRoute(long id, EditRouteDto dto) throws ProviderException {
+        if(!requestUser.hasPermission(UserPermission.TRACK_EDIT))
+            throw new ProviderException(Type.ACCESS_DENIED);
+        DbRoute r = get(DbRoute.class, id, this::isVisible);
+        editFromDto(r, dto);
+        return r;
+    }
+
+    public DbRoute createRoute(EditRouteDto dto) throws ProviderException {
+        if(!requestUser.hasPermission(UserPermission.TRACK_EDIT))
+            throw new ProviderException(Type.ACCESS_DENIED);
+        DbRoute r = new DbRoute();
+        editFromDto(r, dto);
+        em.persist(r);
+        return r;
+    }
+    
+    private void editFromDto(DbRoute r, EditRouteDto dto) throws ProviderException {
+        r.setArchiveAfter(dto.getArchiveAfter());
+        r.setName(dto.getName());
+        r.setForceFirst(dto.getForceFirst());
+        r.setForceLast(dto.getForceLast());
+        r.setLinePoints(dto.getPolyline());
+        r.setOwner(requestUser);
+        r.setTolerance(dto.getTolerance());
+        if(dto.getArchive() != null) {
+            if(r.isArchived() && !dto.getArchive()) {
+                r.setArchived(false);
+                r.setArchiveAfter(0);
+            }
+            if(!r.isArchived() && dto.getArchive()) {
+                r.setArchived(true);
+            }
+        }
+        if(dto.getDeviceId() != null)
+            r.setDevice(devices.getEditableDevice(dto.getDeviceId()));
+        if(dto.getCancel() != null && dto.getCancel()) {
+            r.setStatus(Route.Status.CANCELLED);
+            r.setCancelTimestamp(new Date());
+        }
+        if(r.getCorridor() != null) {
```

Comment:
Jeśli podasz to samo id korytarza, to i tak go usuwasz?
Nie widzę sprawdzenia, czy id się zmieniło.

Reply:
Tak. Tu napotkałem pewien problem, bo w przypadku, gdy zmieni się łamana trasy korytarz staje się nieaktualny. Obsługę zgodności korytarza z łamaną przerzuciłem do API, więc trzeba by sprawdzić, czy obecny korytarz zgadza się z łamaną trasy. Teraz myślę, że raczej chcemy to mieć.